File: java/src/main/java/com/experian/dvs/client/phone/PhoneConfiguration.java
package com.experian.dvs.client.phone;

import com.experian.dvs.client.common.Country;

import java.util.ArrayList;
import java.util.List;

/**
 * AddressConfiguration class for setting up the phone client.
 * Provides options for customizing phone-related API requests.
 */
public class PhoneConfiguration extends com.experian.dvs.client.common.Configuration {

    public static final int DEFAULT_CACHE_VALUE_DAYS = 7;
    private final boolean metadata;
    private final String outputFormat;
    private final int cacheValueDays;
    private final boolean getPortedDate;
    private final boolean getDisposableNumber;
    private final List<Country> liveStatusForMobile;
    private final List<Country> liveStatusForLandline;

    /**
     * Initializes a new instance of the {@link PhoneConfiguration} class using the specified builder.
     *
     * @param builder The builder containing the configuration settings.
     */
    protected PhoneConfiguration(PhoneBuilder builder) {
        super(builder);
        this.metadata = builder.metadata;
        this.outputFormat = builder.outputFormat;
        this.cacheValueDays = builder.cacheValueDays;
        this.getPortedDate = builder.getPortedDate;
        this.getDisposableNumber = builder.getDisposableNumber;
        this.liveStatusForMobile = builder.liveStatusForMobile;
        this.liveStatusForLandline = builder.liveStatusForLandline;
    }

    /**
     * Creates a new instance of the {@link PhoneBuilder} class to configure the phone client.
     *
     * @param token The authentication token for the API.
     * @return A new instance of the {@link PhoneBuilder} class.
     */
    public static PhoneBuilder newBuilder(String token) {
        return new PhoneBuilder(token);
    }

    /**
     * Retrieves the output format setting for the configuration.
     *
     * @return The output format as a string.
     */
    public String getOutputFormat() {
        return outputFormat;
    }

    /**
     * Retrieves the cache value days setting for the configuration.
     *
     * @return The number of days to cache values.
     */
    public int getCacheValueDays() {
        return cacheValueDays;
    }

    /**
     * Retrieves the ported date inclusion setting for the configuration.
     *
     * @return True if the ported date is included; otherwise, false.
     */
    public boolean getGetPortedDate() {
        return getPortedDate;
    }

    /**
     * Retrieves the disposable number inclusion setting for the configuration.
     *
     * @return True if disposable numbers are included; otherwise, false.
     */
    public boolean getGetDisposableNumber() {
        return getDisposableNumber;
    }

    /**
     * Retrieves the list of countries for which live status is enabled for landline numbers.
     *
     * @return A list of {@link Country} objects.
     */
    public List<Country> getLiveStatusForLandline() {
        return liveStatusForLandline;
    }

    /**
     * Retrieves the metadata inclusion setting for the configuration.
     *
     * @return True if metadata is included; otherwise, false.
     */
    public boolean getMetadata() {
        return metadata;
    }

    /**
     * Retrieves the list of countries for which live status is enabled for mobile numbers.
     *
     * @return A list of {@link Country} objects.
     */
    public List<Country> getLiveStatusForMobile() {
        return liveStatusForMobile;
    }

    /**
     * AddressBuilder class for constructing a {@link PhoneConfiguration} object with custom settings.
     */
    public static class PhoneBuilder extends com.experian.dvs.client.common.Configuration.Builder {
        private boolean metadata;
        private String outputFormat = "";
        private int cacheValueDays = DEFAULT_CACHE_VALUE_DAYS;
        private boolean getPortedDate;
        private boolean getDisposableNumber;
        private List<Country> liveStatusForMobile = new ArrayList<Country>();
        private List<Country> liveStatusForLandline = new ArrayList<Country>();

        /**
         * Initializes a new instance of the {@link PhoneBuilder} class with the specified token.
         *
         * @param token The authentication token for the API.
         */
        public PhoneBuilder(String token) {
            super(token);
        }

        /**
         * Sets a custom transaction ID for API requests.
         *
         * @param referenceId The transaction ID to use.
         * @return The current {@link PhoneBuilder} instance for method chaining.
         * @deprecated Set your reference ID as part every API interaction like the Validate call instead.
         */
        @Deprecated(since="1.1.10", forRemoval=true)
        public PhoneBuilder setTransactionId(String referenceId) {
            super.setTransactionId(referenceId);
            return this;
        }

        /**
         * Includes metadata in the API response.
         *
         * @return The current {@link PhoneBuilder} instance for method chaining.
         */
        public PhoneBuilder includeMetadata() {
            this.metadata = true;
            return this;
        }

        /**
         * Sets the output format for the API response.
         *
         * @param outputFormat The output format to use.
         * @return The current {@link PhoneBuilder} instance for method chaining.
         */
        public PhoneBuilder useOutputFormat(String outputFormat) {
            this.outputFormat = outputFormat;
            return this;
        }

        /**
         * Sets the number of days to cache values.
         *
         * @param cacheValueDays The number of days to cache values.
         * @return The current {@link PhoneBuilder} instance for method chaining.
         */
        public PhoneBuilder useCacheValueDays(Integer cacheValueDays) {
            this.cacheValueDays = cacheValueDays;
            return this;
        }

        /**
         * Includes the ported date in the API response.
         *
         * @return The current {@link PhoneBuilder} instance for method chaining.
         */
        public PhoneBuilder includePortedDate() {
            this.getPortedDate = true;
            return this;
        }

        /**
         * Includes disposable numbers in the API response.
         *
         * @return The current {@link PhoneBuilder} instance for method chaining.
         */
        public PhoneBuilder includeDisposableNumber() {
            this.getDisposableNumber = true;
            return this;
        }

        /**
         * Sets the list of countries for which live status is enabled for mobile numbers.
         *
         * @param liveStatusForMobile The list of countries to enable live status for mobile numbers.
         * @return The current {@link PhoneBuilder} instance for method chaining.
         */
        public PhoneBuilder useLiveStatusForMobile(List<Country> liveStatusForMobile) {
            this.liveStatusForMobile = liveStatusForMobile;
            return this;
        }

        /**
         * Sets the list of countries for which live status is enabled for landline numbers.
         *
         * @param liveStatusForLandline The list of countries to enable live status for landline numbers.
         * @return The current {@link PhoneBuilder} instance for method chaining.
         */
        public PhoneBuilder useLiveStatusForLandline(List<Country> liveStatusForLandline) {
            this.liveStatusForLandline = liveStatusForLandline;
            return this;
        }

        /**
         * Builds the {@link PhoneConfiguration} object with the specified settings.
         *
         * @return A new instance of the {@link PhoneConfiguration} class.
         */
        @Override
        public PhoneConfiguration build() {
            return new PhoneConfiguration(this);
        }
    }
}

File: csharp/DVSClient/DVSClient/phone/PhoneConfiguration.cs
﻿using DVSClient.Common;

namespace DVSClient.Phone
{
    /// <summary>
    /// Configuration class for setting up the email validation client.
    /// </summary>
    public class PhoneConfiguration : Configuration
    {
        public const int DefaultCacheValueDays = 7;
        internal bool? Metadata { get; }
        internal string OutputFormat { get; }
        internal int? CacheValueDays { get; }
        internal bool? IncludeDisposableNumber { get; }
        internal bool? IncludePortedDate { get; }
        internal IEnumerable<Country> LiveStatusForMobile { get; }
        internal IEnumerable<Country> LiveStatusForLandline { get; }

        private PhoneConfiguration(PhoneBuilder builder) : base(builder)
        {
            Metadata = builder.Metadata;
            OutputFormat = builder.OutputFormat;
            CacheValueDays = builder.CacheValueDays;
            IncludePortedDate = builder.GetPortedDate;
            IncludeDisposableNumber = builder.GetDisposableNumber;
            LiveStatusForMobile = builder.LiveStatusForMobile;
            LiveStatusForLandline = builder.LiveStatusForLandline;
        }

        /// <summary>
        /// Creates a new builder instance for configuring the phone validation client.
        /// </summary>
        /// <param name="token">The API token to authenticate requests.</param>
        /// <returns>A new instance of the Builder class.</returns>
        public static PhoneBuilder NewBuilder(string? token)
        {
            return new PhoneBuilder(token);
        }

        public class PhoneBuilder : Builder
        {
            internal bool Metadata { get; private set; }
            internal string OutputFormat { get; private set; } = string.Empty;
            internal int CacheValueDays { get; private set; } = DefaultCacheValueDays;
            internal bool GetPortedDate { get; private set; }
            internal bool GetDisposableNumber { get; private set; }
            internal IEnumerable<Country> LiveStatusForMobile { get; private set; } = new List<Country>();
            internal IEnumerable<Country> LiveStatusForLandline { get; private set; } = new List<Country>();

            /// <summary>
            /// Initializes a new instance of the Builder class.
            /// </summary>
            /// <param name="token">The API token to authenticate requests.</param>
            public PhoneBuilder(string? token) : base(token)
            {
            }

            /// <summary>
            /// Sets the maximum delay between retries for API requests.
            /// </summary>
            /// <param name="maxDelay">The maximum delay in milliseconds.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new PhoneBuilder SetMaxDelay(int maxDelay)
            {
                base.SetMaxDelay(maxDelay);
                return this;
            }

            /// <summary>
            /// Sets the initial delay before retrying API requests.
            /// </summary>
            /// <param name="initialDelay">The initial delay in milliseconds.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new PhoneBuilder SetInitialDelay(int initialDelay)
            {
                base.SetInitialDelay(initialDelay);
                return this;
            }

            /// <summary>
            /// Sets the maximum number of retries for API requests.
            /// </summary>
            /// <param name="maxRetries">The maximum number of retries.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new PhoneBuilder SetMaxRetries(int maxRetries)
            {
                base.SetMaxRetries(maxRetries);
                return this;
            }

            /// <summary>
            /// Sets the timeout for API requests in seconds.
            /// </summary>
            /// <param name="timeouts">The timeout duration in seconds.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new PhoneBuilder SetApiRequestTimeoutInSeconds(int timeouts)
            {
                base.SetApiRequestTimeoutInSeconds(timeouts);
                return this;
            }

            /// <summary>
            /// Sets the timeout for the HTTP client in seconds.
            /// </summary>
            /// <param name="timeouts">The timeout duration in seconds.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public new PhoneBuilder SetHttpClientTimeoutInSeconds(int timeouts)
            {
                base.SetHttpClientTimeoutInSeconds(timeouts);
                return this;
            }

            /// <summary>
            /// Sets a custom reference ID for API requests.
            /// </summary>
            /// <param name="referenceId">The reference ID to use.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            [Obsolete("Set your reference ID as part every API interaction like the Validate call instead.")]
            public new PhoneBuilder SetTransactionId(string referenceId)
            {
                base.SetTransactionId(referenceId);
                return this;
            }

            /// <summary>
            /// Includes metadata in the API response.
            /// </summary>
            /// <returns>The current Builder instance for method chaining.</returns>
            public PhoneBuilder IncludeMetadata()
            {
                Metadata = true;
                return this;
            }

            /// <summary>
            /// Specifies the phone number output format for the API response.
            /// </summary>
            /// <param name="outputFormat">The desired output format (e.g., NATIONAL).</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public PhoneBuilder UseOutputFormat(string outputFormat)
            {
                OutputFormat = outputFormat;
                return this;
            }

            /// <summary>
            /// Sets the maximum ages of the cache for the phone number validation result.
            /// </summary>
            /// <param name="cacheValueDays">The maximum age of the cache.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public PhoneBuilder UseCacheValueDays(int cacheValueDays)
            {
                CacheValueDays = cacheValueDays;
                return this;
            }

            /// <summary>
            /// Includes the ported date of the phone number in the API response.
            /// </summary>
            /// <returns>The current Builder instance for method chaining.</returns>
            public PhoneBuilder IncludePortedDate()
            {
                GetPortedDate = true;
                return this;
            }

            /// <summary>
            /// Includes information about whether the phone number is disposable in the API response.
            /// </summary>
            /// <returns>The current Builder instance for method chaining.</returns>
            public PhoneBuilder IncludeDisposableNumber()
            {
                GetDisposableNumber = true;
                return this;
            }

            /// <summary>
            /// Specifies the countries for which live status should be checked for mobile numbers.
            /// </summary>
            /// <param name="liveStatusForMobile">A list of countries to check live status for mobile numbers.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public PhoneBuilder UseLiveStatusForMobile(IEnumerable<Country> liveStatusForMobile)
            {
                LiveStatusForMobile = liveStatusForMobile;
                return this;
            }

            /// <summary>
            /// Specifies the countries for which live status should be checked for landline numbers.
            /// </summary>
            /// <param name="liveStatusForLandline">A list of countries to check live status for landline numbers.</param>
            /// <returns>The current Builder instance for method chaining.</returns>
            public PhoneBuilder UseLiveStatusForLandline(IEnumerable<Country> liveStatusForLandline)
            {
                LiveStatusForLandline = liveStatusForLandline;
                return this;
            }

            /// <summary>
            /// Builds the Configuration object with the specified settings.
            /// </summary>
            /// <returns>A Configuration object with the configured settings.</returns>
            public override PhoneConfiguration Build()
            {
                return new PhoneConfiguration(this);
            }
        }
    }
}
